### PR TITLE
JAVA-1494: Implement Snappy and LZ4 compression

### PR DESCRIPTION
**Pull request by olim7t**
This was scheduled for alpha3 so no pressure, but if time permits we could include it in alpha2.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -50,7 +50,7 @@
       <dependency>
         <groupId>com.datastax.oss</groupId>
         <artifactId>native-protocol</artifactId>
-        <version>1.4.0</version>
+        <version>1.4.1-SNAPSHOT</version>
```
Comment (olim7t) [inline pom.xml:53]: This requires a local install of native-protocol. I'll do the release at the last minute in case other changes come up.

File: core/src/main/java/com/datastax/oss/driver/internal/core/protocol/ByteBufCompressor.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.datastax.oss.protocol.internal.Compressor;
+import io.netty.buffer.ByteBuf;
+import java.nio.ByteBuffer;
+
+public abstract class ByteBufCompressor implements Compressor<ByteBuf> {
+
+  @Override
+  public ByteBuf compress(ByteBuf uncompressed) {
+    return uncompressed.isDirect() ? compressDirect(uncompressed) : compressHeap(uncompressed);
```
Comment (tolbertam) [inline core/src/main/java/com/datastax/oss/driver/internal/core/protocol/ByteBufCompressor.java:26]: Nice to break up the direct/heap implementations, that made the code a bit hard to follow on 3.x. 👍

**Comment by tolbertam** [comment]
I had to go back to `HEAD~2` on native-protocol to get a snapshot that works (I think recent changes on native-protocol aren't compatible with this branch).

File: integration-tests/src/test/java/com/datastax/oss/driver/api/core/compression/DirectCompressionIT.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  public void should_execute_queries_with_lz4_compression() throws Exception {
+    createAndCheckCluster(
+        "protocol.compressor.class = "
+            + "com.datastax.oss.driver.internal.core.protocol.Lz4Compressor");
+  }
+
+  private void createAndCheckCluster(String compressorOption) {
+
+    try (Cluster cluster =
+        ClusterUtils.newCluster(ccmRule, compressorOption, "request.timeout = 30 seconds")) {
```
Comment (tolbertam) [inline integration-tests/src/test/java/com/datastax/oss/driver/api/core/compression/DirectCompressionIT.java]: The request.timeout isn't needed here right?  That's more needed for making schema changes I had thought.

**Review by tolbertam — COMMENTED** [review]
Looks great so far, the tests follow what was done on 3.x so I consider that adequate.  I'm doing some quick local testing with large throughput to see if the memory profiles look like what I'm used to seeing with direct/heap use and then will approve.  I pushed a small change to tests to remove the additional use of larger timeout which I don't think is needed.

**Comment by tolbertam** [comment]
Unsure if this is related to this PR, but i noticed that the netty leak detection fired off when running with Snappy and LZ4 compression ~(not sure if isolated to snappy, checking lz4)~:

```
ERROR [2017-10-13 10:15:22,489] [c0-io-13            ] io.netty.util.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected. Enable advanced leak reporting to find out where the leak occurred. To enable advanced leak reporting, specify the JVM option '-Dio.netty.leakDetection.level=advanced' or call ResourceLeakDetector.setLevel() See http://netty.io/wiki/reference-counted-objects.html for more information.
```

Will turn on advanced detection and see what I find.

Note that the warning does not pop up if compression is not enabled.

**Comment by tolbertam** [comment]
Leak detection output:  https://gist.github.com/tolbertam/1269be8a4efd899d1a82220e2637fee8

**Comment by tolbertam** [comment]
Wonder if it's because we're not releasing `compressedMessage` in the native-protocol code in `FrameCodec.encode`?  Will try it out

```
      // 2) Compress and measure size, discard uncompressed buffer
      B compressedMessage = compressor.compress(uncompressedMessage);
      primitiveCodec.release(uncompressedMessage);
      int messageSize = primitiveCodec.sizeOf(compressedMessage);

      // 3) Encode final frame
      B header = primitiveCodec.allocate(headerSize);
      encodeHeader(frame, flags, messageSize, header);
      return primitiveCodec.concat(header, compressedMessage);
```

**Comment by tolbertam** [comment]
I was on the wrong track, the problem is in decompress, not compress.  Looking into it.

**Comment by tolbertam** [comment]
I think I have a handle on this, simply just needed to add a `primitiveCodec.release(source)` in `FrameCodec.decode` and the tracker no longer detects any leaks:

```java

    if (flags.contains(Flag.COMPRESSED)) {
      source = compressor.decompress(source);
    }

    boolean isTracing = flags.contains(Flag.TRACING);
    UUID tracingId = (isResponse && isTracing) ? primitiveCodec.readUuid(source) : null;

    Map<String, ByteBuffer> customPayload =
        (flags.contains(Flag.CUSTOM_PAYLOAD))
            ? primitiveCodec.readBytesMap(source)
            : Collections.emptyMap();

    List<String> warnings =
        (isResponse && flags.contains(Flag.WARNING))
            ? primitiveCodec.readStringList(source)
            : Collections.emptyList();

    Message.Codec decoder = decoders.get(protocolVersion, opcode);
    ProtocolErrors.check(
        decoder != null, "Unsupported request opcode: %s in protocol %d", opcode, protocolVersion);
    Message response = decoder.decode(source, primitiveCodec);

    primitiveCodec.release(source); // <----HERE

    return new Frame(
        protocolVersion, beta, streamId, isTracing, tracingId, customPayload, warnings, response);
```

However, this causes issues with compression disabled, (source is released upstream otherwise).  So maybe we should change the code to only release if the reference returned by `decompress` is different than the input.

**Review by tolbertam — APPROVED** [review]
Working great 👍 .  Ran ~2 minute duration tests with uncompressed, Snappy and LZ4 compressors in both direct and heap mode.  Everything worked as detected and profile showed correct paths being used with no leaks detected with advanced detection enabled.
